### ISSUE TITLE
Add support for extracting a bit mask from a Vector<T>

### ISSUE THREAD

**Issue by mburbea** [dotnet/corefx#1010]
SSE2/AVX supports extracting the MSB from each 8 byte,4 byte,2 byte or 1 byte segment (the `movemask_*` family). This can be extremely useful while performing algorithms that are data sensitive and use SSE registers like big buffers. 

A user could implement algorithms that work entirely in safe code that could benefit from avoiding the allocation and rigmarole of copying data out of a register to an array when you just want to test if a field is not empty. 

Admittedly, it is a downer as ARM does not have an intrinsic for this operation but there is workarounds that should perform better than copying everything in and out memory to test the bytes.
http://stackoverflow.com/questions/11870910/sse-mm-movemask-epi8-equivalent-method-for-arm-neon

**Comment by mellinoe** [comment]
This could be a very interesting addition. Could you give some examples of how you would expect the operations to work, i.e what would the return type of various operations be, how could you use them, etc. Especially with regards to different data types and different Vector<T> sizes (from various hardware capabilities).

The closest thing we have right now is ConditionalSelect, which can be used to select between two vectors based on a third "selection" mask vector. Masking operations could probably just be seen as a special case of this operation, but I think extracting a bit mask based on the MSB of each element could be useful as well, to then be used as a mask for ConditionalSelect.

**Comment by mburbea** [comment]
Well if this gets considered, I would probably then request a `BitScanForward` or `BitScanBackward` method on `int`/`uint`/`long`/`ulong`. Maybe on a bitvector Otherwise your forced to use masks to try to find the set values. 

Will `Vector<T>` allow variable register size allocation? E.g. if you are reading through a `byte[]` and there are only 16 slots left but have a processor with 256 bit register support, could I allocate a SSE2 register?

 Regardless of vector size it should always return a `ulong` or if a `BitScanForward` instrinsic will not be added maybe a `BitVector64` (a new struct similar to the old `BitVector32`). This will handle cases of extracting bit masks on bytes on vector sizes of up to 512bit without needing to break out into multiple values. 

I think the method should be based on the generic type of the vector e.g. For `double`,`long` and `ulong` it should only call ` _mm_movemask_pd`. And only the first 2 bits would be relevent in SSE2/NEON, AVX256/512 would have either 4 or 8 relevant bits.  With some bittwiddling before extracting the mask you could use this to determine which elements have a value.

**Comment by mellinoe** [comment]
I still believe this would be a useful addition, but like many of the others for `Vector<T>`, we need to come up with a concrete API proposal to make it actionable.

@mburbea I know this issue is very, very old, but let me know if you have any further thoughts about this, or concrete ideas for API additions.

**Comment by mburbea** [comment]
I was thinking about it.
I think the following methods could be beneficial
```
Vector<T>{
long GetBitsSet()
ulong GetBitSetUnsigned()
bool IsZero(int Pos)
}
```
Regardless, of the `T`, it should always return a long. GetBitSet will call the appropriate mask method and return the mask. Since VB code cannot use non-cls complaint types both methods would be needed. 

IsZero would be great as the jit can just extract the mask and check if the bit is zero. I find IsZero a more comfortable name but I feel most people will use it as `!myVector.IsZero(i)`, so maybe a better name could be in order
Repeated ones in a loop could be optimized to only extract the mask once and visit it over and over again.

**Comment by mellinoe** [comment]
The first two overloads seem reasonable; the last I'm not sure about. It sounds like it would basically just call `GetBitMask` and then check if the `Pos`-th digit in the return value was 0? We've talked about some sort of `BitManipulation` class in the past, the last function could be implemented something like:

```CSharp
public static class BitManipulation
{
    public static bool IsBitZero(long value, int bitIndex);
}
```

I'm only mentioning the above because it feels like `IsZero` is actually just a particular use-case of a more general function.

The other concern I'd have is that the `long` return type could be problematic in theoretical future architectures where register sizes exceed 512 bits. Byte vectors could then potentially have greater than 64 elements, and mask elements would therefore also exceed 64. Thoughts on that part?

**Comment by mburbea** [comment]
I agree but I think it somewhat generalizes a use case.  A bit manipulation class could be a nice way to expose intrinsics like BitScanForward/BitScanReverse.

As for the concern of growing vectors, there could be a few solutions.
```csharp
Vector<T>{
public int GetBitMaskInt32(int offset)
public uint GetBitMaskUInt32(int offset)
public int GetBitMaskInt64(int offset)
public uint GetBitMaskUInt64(int offset)
}
```
Essentially, allow you to extract the bit mask offset from the `offset`-th slot of your type. For the current implementation, everyone could just use `ExtractBitMaskUInt32(0)`. Requesting something like `ExtractBitMaskInt32(2)` on `Vector<long>` would throw an out of range exception.
It's not the best I'll admit.
The other way could be a hardware specific struct.
```Vector<T>{
public BitMask GetBitMask()
}

public struct BitMask
{
     bool this[int bit]
     public int FindFirstNonZeroBit()
     public int FindLastNonZeroBit()
     GetInt32(int offset)
     GetUInt32(int offset)
     GetInt64(int offset)
     GetUInt64(uint offset)
     ...
}
```
So, when Avx1024 rolls around and the k registers become 128bit types (maybe they'll be vector types themselves who knows), the jit can emit an effecient call to `_mm1024_movepi8_mask` or whatever it's new intrinsic will be and then this type can be specialized for operations one might want to do on the mask.

**Comment by mellinoe** [comment]
@mburbea Thanks for the proposal. I think that this is a good starting point:

```CSharp
public int GetBitMaskInt32(int offset)
public uint GetBitMaskUInt32(int offset)
public int GetBitMaskInt64(int offset)
public uint GetBitMaskUInt64(int offset)
```

We could also consider just starting with "parameterless" versions for now, which would be the equivalent of `GetBitMaskX(0)`.

@CarolEidt How does the above sound?

Right now we have quite a few items on our backlog for `Vector<T>`, so I can't really say that we will have the spare cycles to build and test this particular addition any time soon. But from a design perspective, I do think it looks good.

**Comment by danmosemsft** [comment]
@CarolEidt for the question above

**Comment by CarolEidt** [comment]
@danmosemsft - I'm a little confused as to what the offset means. Does that mean to return the low-order {32 or 64} bits of the bit mask containing the high bits of the elements in Vector<T> starting from the offset'th element? It might be good if we had some pseudo code for this just to make it clear.
And I should second what @mellinoe says - I'm juggling a few things right now, and am probably less than half done with the implementation of the Convert/Narrow/Widen intrinsics. So it might be a while ...
But the sooner we get to resolution on these the better, as they're the kinds of things that I can work on in spare cycles.

**Comment by mellinoe** [comment]
> I'm a little confused as to what the offset means. Does that mean to return the low-order {32 or 64} bits of the bit mask containing the high bits of the elements in Vector starting from the offset'th element?

Yes, that is correct. It would be for cases where vectors have greater than 32/64 elements in them. We could consider just exposing the parameterless version at first, that one being equivalent to passing `0` in.

**Comment by danmosemsft** [comment]
@CarolEidt community are implementing APIs faster than we can approve them right now. I'm trying to speed up getting some of these old API proposals to the review meeting to make sure they don't run out of work...
